stp_zmq/test/test_stashed_client_messages.py

```diff
@@ -0,0 +1,190 @@
+    msg = {'msg': 'msg1' * tconf.MSG_LEN_LIMIT}
+    assert not pending_client_messages
+    assert not alpha_handler.received_messages
+
+    alpha.connect(name=beta.name, ha=beta.ha,
+                  verKeyRaw=beta.verKeyRaw, publicKeyRaw=beta.publicKeyRaw)
+
+    looper.runFor(0.25)
+
+    alpha.send(msg, beta.name)
+    assert not pending_client_messages
+    assert not alpha_handler.received_messages
+
+    msg = {'msg': 'msg1'}
+    alpha.send(msg, beta.name)
+    looper.run(eventually(
+        lambda msg_handler: assertExp(msg_handler.received_messages == [msg]),
+        alpha_handler))
+
+
+def test_resending_for_old_stash_msgs(tdir, tconf, looper, stacks,
+                                      alpha_handler, monkeypatch):
+    alpha, beta = stacks
+    msg1 = {'msg': 'msg1'}
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+
+    alpha.connect(name=beta.name, ha=beta.ha,
+                  verKeyRaw=beta.verKeyRaw, publicKeyRaw=beta.publicKeyRaw)
+    looper.runFor(0.25)
+
+    def fake_send_multipart(msg_parts, flags=0, copy=True, track=False, **kwargs):
+        raise zmq.Again
+
+    monkeypatch.setattr(beta.listener, 'send_multipart',
+                        fake_send_multipart)
+    alpha.send(msg1, beta.name)
+    looper.run(eventually(
+        lambda messages: assertExp(messages[alpha.listener.IDENTITY] == [(0, msg1)]),
+        pending_client_messages))
+    monkeypatch.undo()
+
+    beta._client_message_provider._timer.set_time(tconf.RESEND_CLIENT_MSG_TIMEOUT + 2)
+    looper.run(eventually(
+        lambda msg_handler: assertExp(msg_handler.received_messages == [msg1]),
+        alpha_handler))
+    assert not pending_client_messages
+
+
+def test_limit_msgs_for_client(tconf, looper, stacks, alpha_handler):
+    alpha, beta = stacks
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+    for i in range(tconf.PENDING_MESSAGES_FOR_ONE_CLIENT_LIMIT + 2):
+        beta.send(create_msg(i),
+                  alpha.listener.IDENTITY)
+    assert len(pending_client_messages[alpha.listener.IDENTITY]) == tconf.PENDING_MESSAGES_FOR_ONE_CLIENT_LIMIT
+    assert pending_client_messages[alpha.listener.IDENTITY][0] != (0, create_msg(0))
+    assert pending_client_messages[alpha.listener.IDENTITY][-1] == (0,
+                                                                    create_msg(tconf.PENDING_MESSAGES_FOR_ONE_CLIENT_LIMIT + 1))
+
+
+def test_limit_pending_queue(tconf, looper, stacks, alpha_handler):
+    alpha, beta = stacks
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+    for i in range(tconf.PENDING_CLIENT_MESSAGES_LIMIT + 2):
+        beta.send(create_msg(1), str(i))
+    assert len(pending_client_messages) == tconf.PENDING_CLIENT_MESSAGES_LIMIT
+    assert str(0) not in pending_client_messages
+    assert str(tconf.PENDING_CLIENT_MESSAGES_LIMIT + 1) in pending_client_messages
+
+
```

Comment:
I think the following test can be added in addition: test that only the pending messages for the given client are sent with the next `send` call
- have pending messages for 2 clients
- connect one of the client and send a message to this
- make sure that the pending message to this client is sent, but pending message to the second client isn't